src/learner.cc

```diff
@@ -151,7 +152,7 @@ DMLC_REGISTER_PARAMETER(LearnerTrainParam);
  * \brief learner that performs gradient boosting for a specific objective
  * function. It does training and prediction.
  */
-class LearnerImpl : public Learner {
+class LearnerImpl : public Learner, public LearnerTestHook {
```

Comment:
This feels weird to me.  If I were you I would first add `std::map<std::string, std::string> Learner::GetConfiguration() const` abstract method to `Learner`, let `LearnerImpl` to override it.  Then call `Learner::Create()` to make a learner pointer in testing code, which can be used to obtain the configuration you are interested.

But that's just personal preference.

Reply:
Another option would be declare `LearnerImpl` in a header, then create a `LearnerMock` in testing code, which inherits `Learner`.  Then you can use `LearnerMock` to access the internal of `LearnerImpl`.

Reply:
@trivialfis The options you presented are all valid. I chose this approach so that the change to `Learner` and `LearnerImpl` classes is minimized. Also, it would be easy to add access (from the test) to more internal stuff in Learner.